Source/WebCore/html/FormAssociatedCustomElement.cpp

```diff
@@ -0,0 +1,209 @@
+            WTF::switchOn(item.data, [&](const String& value) {
+                formData.append(item.name, value);
+            }, [&](RefPtr<File> value) {
+                formData.append(item.name, *value);
+            });
+        }
+    }, [&](const String& value) {
+        if (!name().isEmpty())
+            formData.append(name(), value);
+    }, [&](RefPtr<File> value) {
+        if (!name().isEmpty())
+            formData.append(name(), *value);
+    });
+
+    return true;
+}
+
+void FormAssociatedCustomElement::formWillBeDestroyed()
+{
+    m_formWillBeDestroyed = true;
+    ValidatedFormListedElement::formWillBeDestroyed();
+    m_formWillBeDestroyed = false;
+}
+
+void FormAssociatedCustomElement::reset()
+{
+    if (m_element->canAccessFormAssociatedInternals())
+        CustomElementReactionQueue::enqueueFormResetCallbackIfNeeded(*m_element);
+}
+
+void FormAssociatedCustomElement::disabledStateChanged()
+{
+    ValidatedFormListedElement::disabledStateChanged();
+
+    if (m_element->canAccessFormAssociatedInternals())
+        CustomElementReactionQueue::enqueueFormDisabledCallbackIfNeeded(*m_element, isDisabled());
+}
+
+void FormAssociatedCustomElement::didChangeForm()
+{
+    ValidatedFormListedElement::didChangeForm();
+
+    if (m_element->canAccessFormAssociatedInternals() && !m_formWillBeDestroyed)
+        CustomElementReactionQueue::enqueueFormAssociatedCallbackIfNeeded(*m_element, form());
+}
+
+void FormAssociatedCustomElement::didUpgrade(JSC::JSGlobalObject& lexicalGlobalObject)
+{
+    CustomElementReactionStack customElementReactionStack(lexicalGlobalObject);
+
+    HTMLElement& element = asHTMLElement();
+
+    parseFormAttribute(element.attributeWithoutSynchronization(formAttr));
+    parseDisabledAttribute(element.attributeWithoutSynchronization(disabledAttr));
+    parseReadonlyAttribute(element.attributeWithoutSynchronization(readonlyAttr));
+
+    resetDataListAncestorState();
+    updateWillValidateAndValidity();
+    syncWithFieldsetAncestors(element.parentNode());
+    invalidateElementsCollectionCachesInAncestors();
+}
+
+void FormAssociatedCustomElement::invalidateElementsCollectionCachesInAncestors()
+{
+    auto invalidateElementsCache = [](HTMLElement& element) {
+        if (auto* nodeLists = element.nodeLists())
+            nodeLists->invalidateCachesForAttribute(HTMLNames::formAttr);
+    };
+
+    if (RefPtr<HTMLFormElement> form = this->form())
+        invalidateElementsCache(*form);
+
+    for (auto& ancestor : lineageOfType<HTMLFieldSetElement>(*m_element))
+        invalidateElementsCache(ancestor);
+}
+
+const AtomString& FormAssociatedCustomElement::formControlType() const
+{
+    return asHTMLElement().localName();
+}
+
+FormControlState FormAssociatedCustomElement::saveFormControlState() const
+{
+    FormControlState savedState;
+
+    if (m_state.has_value() && std::holds_alternative<String>(m_state.value()))
+        savedState.append(std::get<String>(m_state.value()));
```

Comment:
Maybe add FIXME to support other types?

Reply:
Added support for string-only `FormData` in 3f7937b and also put a FIXME regarding still missing `File` type support.